Remove hash binding for ML-DSA hybrid signatures (#166)
openpgp-pqc/draft-openpgp-pqc: draft-ietf-openpgp-pqc.md

## Composite Signature Schemes with ML-DSA {#ecc-mldsa}

> 

Author: ```suggestion

### Signature data digest {#mldsa-sig-data-digest}
In order not to extend the attack surface, composite ML-DSA + EdDSA signatures SHOULD use SHA3 for the signature data digest because ML-DSA internally uses a SHAKE256 digest.
Implementations MAY allow use of SHA2 for CNSA 2.0 compliance.
```

## Composite Signature Schemes with ML-DSA {#ecc-mldsa}

### Signature data digest {#mldsa-sig-data-digest}

> In order not to extend the attack surface, composite ML-DSA + EdDSA signatures SHOULD use SHA3 for the signature data digest because ML-DSA internally uses a SHAKE256 digest.

Reviewer: SHA2-384 and SHA2-512 are internationally approved hash functions and therefore a viable choice to compute a message digest. We should not claim or hint at, that the use of those SHA2 instances for the signature data digest extends the attack surface.

Reviewer 2: It's not about security concerns with SHA2 specifically but about extending the attack surface due to adding an additional hash algorithm from a different algorithm family to the signature operation (if the digest algorithm doesn't match the internally used algorithm). 

Maybe it needs better wording?

### Signature data digest {#mldsa-sig-data-digest}

Signature data (i.e. the data to be signed) is digested prior to signing operations, see [[RFC9580, Section 5.2.4]](https://www.rfc-editor.org/rfc/rfc9580#section-5.2.4).

> Composite ML-DSA + EdDSA signatures MUST use an associated hash algorithm as specified in {{tab-mldsa-hash}} for the signature data digest. They SHOULD use SHA3 to match the ML-DSA internal use of SHAKE256. They MAY use SHA2 for CNSA 2.0 compliance. Signatures using other combinations or hash algorithms MUST be considered invalid.

Reviewer 3: I would not write anything about CNSA. It we allow SHA-2, we allow it, I don't think it makes sense to put that under further requirements (otherwise it ends up in a weird mixture of conditional MAY and MUST requirements).

 I would write:
> They MAY use SHA2. This can for instance be necessary to satisfy specific regulatory requirements.

> Composite ML-DSA + EdDSA signatures MUST use an associated hash algorithm as specified in {{tab-mldsa-hash}} for the signature data digest. They SHOULD use SHA3 to match the ML-DSA internal use of SHAKE256. They MAY use SHA2 for CNSA 2.0 compliance. Signatures using other combinations or hash algorithms MUST be considered invalid.

Author: ```suggestion
Composite ML-DSA + EdDSA signatures MUST use an associated hash algorithm as specified in {{tab-mldsa-hash}} for the signature data digest. They SHOULD use SHA3 to match the ML-DSA internal use of SHAKE256. They MAY use SHA2, this can for instance be necessary to satisfy specific regulatory requirements. Signatures using other combinations or hash algorithms MUST be considered invalid.
```

This specification makes use of the default "hedged" variants of ML-DSA and SLH-DSA, which mix fresh randomness into the respective signature-generation algorithm's internal hashing step.
This has the advantage of an enhanced side-channel resistance of the signature operations according to  {{FIPS-204}} and {{FIPS-205}}.

## Binding data digests and signature algorithms

Signature data (i.e. the data to be signed) is digested prior to signing operations, see [[RFC9580, Section 5.2.4]](https://www.rfc-editor.org/rfc/rfc9580#section-5.2.4).

> Binding a NIST security category 2 hash function (e.g. SHA3-256) to a signature algorithm that falls into NIST security category 3 or 5 does not constitute a security bottleneck: because of the unpredictable random salt that is prepended to the digested data in v6 signatures, the hardness assumption is not collision resistance but second-preimage resistance.

Reviewer 2: 1. I would not use the term "binding" any more.
2. Since we do not change anything in the OpenPGP protocol we do not need to reiterate that signature data is hashed before signing, At least the reference to RFC 9580 is overkill IMO.
3. The section seems to be about NIST security categories and how the salt helps to achieve higher security levels with "weaker" hash functions. There is no reference or introduction to the NIST security categories. I think that would be needed.

Suggestion: Rework the section and title, make it about what NIST security categories are achieved.

Alternative suggestion: Just remove it entirely. I think the benefit of the salt regarding collision / second preimage resistance is already part of RFC 9580. I'm not sure how relevant the NIST security categories are for this document.

Third alternative: Just add a one-sentence statement about the salt and second-preimage resistance somewhere.

Reviewer: I just boiled down already existing text. But I see that it does not add that much anymore, now that we have removed the bindings. I would go with the alternative suggestion and remove the section completely, so that we simply add new signature algorithms build upon RFC9580.

This specification makes use of the default "hedged" variants of ML-DSA and SLH-DSA, which mix fresh randomness into the respective signature-generation algorithm's internal hashing step.
This has the advantage of an enhanced side-channel resistance of the signature operations according to  {{FIPS-204}} and {{FIPS-205}}.

## Binding digest hash in signatures with SLH-DSA

In order not to extend the attack surface, we bind the hash algorithm used for signature data digestion to the hash algorithm used internally by the SLH-DSA signature algorithm, see {{slhdsa-sig-data-digest}}.

> This is not done for ML-DSA to comply with the CNSA 2.0 guideline.

Reviewer 2: I would not argue for ML-DSA in the SLH-DSA section. I would either move this to its own section or rename the title s.t. it does not exclude ML-DSA.

Author: ```suggestion
```

- Fixed and improved test vectors.

## draft-ietf-openpgp-pqc-07
- Assigned code points 30 - 34 for ML-DSA and SLH-DSA algorithms.
- Aligned KEM combiner with LAMPS
- Dropped CCA-conversion of X25519/X448
- Switched to hedged variant also for SLH-DSA

## draft-ietf-openpgp-pqc-08

> - Removed hash binding for ML-DSA signatures

Author: ```suggestion
- Changed hash binding for ML-DSA signatures
```
